### ISSUE TITLE
un-depreciate scher multiplication in vector.multiply() and add to vector metatables 

### ISSUE THREAD

**Issue by FatalError42O** [minetest/minetest#12533]
## Problem

There's absolutely no need to have scher multiplication in vector.multiply() deprecated- it's a useful tool for modders, it's especially useful for things like scaling. I don't see the need to discontinue it, it's unnecessary and fundamentally counter productive, as when I multiply two vectors I expect this product.

Furthermore, it needs to be implemented in metatable vector multiplication (i.e. `vector2*vector1`) 

## Solutions

Un-depreciate `vector.multiply(vector1, vector2)`
implement Scher for metatables such as `vector1*vector2`

## Alternatives

Make a alternative function for this, however I don't see the point in this, it's rather redundant and wasteful, moreover there's no way to represent it with the new arithmetic on vectors.

## Additional context

Once again, I don't see the reason this was ever unimplemented, it's incredibly useful... there's never a good reason to simply remove or discontinue a feature just because you don't like it.

**Comment by appgurueu** [comment]
> There's absolutely no need to have Schur multiplication in vector.multiply() deprecated- it's a useful tool for modders, it's especially useful for things like scaling.

100% agreed, the deprecation **without alternative** was a mistake. Usually functions are deprecated in favor of a cleaner alternative. This was not the case here, the functionality was wrongly considered "useless" and thus marked for removal.

> Make an alternative function for this, however I don't see the point in this, it's rather redundant and wasteful, moreover there's no way to represent it with the new arithmetic on vectors.

"Redundant" and "wasteful"? Not at all. It's rather **explicit**. For readability the purposes, the new function could be called `vector.scale` or `vector.schur_product` (probably the former to make it convenient for beginners).

> Furthermore, it needs to be implemented in metatable vector multiplication (i.e. `vector2*vector1`)

Hard disagreement here: Multiplication of vectors denotes the dot product in maths (hence why it's called the *dot* product). `v*w` should be a convenience form of `vector.dot(v, w)`, which is used more frequently than the Schur product. As `vector.multiply(v, w)` should ideally be the same as `v * w`, it makes sense to deprecate `vector.multiply(v, w)` being the Schur product. **An alternative function should have been provided however.** Simply removing useful functionality does not help the engine.

Side note: Since we now have `vector.combine`, you could use that for the Schur product: `vector.combine(v, w, function(a, b) return a*b end)`.

**Comment by Desour** [comment]
No objection from my side.
You're right, this deprecation was probably a bad idea, and not justified.

------------------------

>Hard disagreement here: Multiplication of vectors denotes the dot product in maths (hence why it's called the dot product).

Actually, I most often see the dot product written as `<v1,v2>`. But I'm not sure if that's a german thing.

Instead of looking at math notation, we could also look at what other libraries and co. do. I.e. glsl does schur product with `*`.

**Comment by erlehmann** [comment]
> Perhaps we should in the end go with @erlehmann's suggestion of forcing the user to make it explicit by allowing neither.

FWIW, I actually asked a mathematician working in IT what he would think `v * w` means and  he first asked if I had meant `v × w` or `v • w`. When I told him that the multiplication character is `*`, he was immediately in favor of forcing everyone to write it out to avoid surprises.

**Comment by Montandalar** [comment]
I agree with erlehmann's friend. Please make people write out what they mean. * is not × or •. However many keystrokes people will save by typing * are not outweighed by the fact that many people casually reading the code may misunderstand it.

**Comment by Desour** [comment]
To what conclusion do we come here? We could:
(1) Do component wise stuff (aka the glsl way): Allow `v1*v2`, `v+s`, `s+v` (maybe even `vector.multiply(s, v)`?).
(2) Add `.schur`: Add `vector.schur(v1,v2)`, keep `vector.multiply(v1,v2)` deprecated. We should also deprecate `vector.add(v, s)` in this case, with the same reasoning, and replace it with something else, like `vector.offset_all(s)`.
(>=3) Something else?

I'd prefer option (1), as `v1*v2` is not *really* confusing anyone (if one had to guess what it does, most would guess schur product first, I assume). (2) would complicate the interface for the user, and I don't see any substantial benefits.

**Comment by erlehmann** [comment]
@FatalError42O @appgurueu I am looking at this issue right now since the schur product has come up in a semi-private conversation right now. So, is there a PR for undeprecation … or anything?

**Comment by appgurueu** [comment]
In hindsight, I find my yapping here silly.

Let's just undeprecate it. Probably the most simple solution that should make almost everyone happy. There's ample precedent for this dating back to the APL days. It's commonly found in vector / ndarray libs, e.g. numpy, GLSL.

Component-wise notation may not be that common in mathematics, but it is common in array programming (graphics programming / gamedev), and often convenient. I doubt the "explicit-ness" adds much value.

Side note: For consistency, both Schur product and Schur quotient should be undeprecated.
